devdoc/Metrics.rst

```diff
@@ -0,0 +1,282 @@
+
+ESP needs to report the *activeTransactions* value from the TxSummary class(es). This gives an indication of how busy the ESP is in terms of client requests.
+
+Direct measurement of response time in requests may not be useful since the type of request causes different execution paths within ESP that are expected to take widely varying amounts of time. Creation of metrics for each method is not recommended. However, two possible solutions are to a) create a metric for request types, or b) use a histogram to measure response time ranges. Another option mentioned redefines the meaning of a bucket in a histogram. Instead of a numeric distribution, each bucket represents a unique subtask within an overall "metric" representing a measured operation. This should be explored whether for operational or developmental purposes.
+
+For tracking specific queries and their health, the feeling is that logging can accomplish this better than metrics since the list of queries to monitor will vary between clusters. Additionally, operational metrics solving the cases mentioned above will give a view into the overall health of ESP which will affect the execution of queries. Depending on actions taken by these metrics, scaling may solve overload conditions to keep cluster responsiveness acceptable. Internal problems with queries should be diagnosed from deep drill down logging.
+
+For Roxie a workunit operates as a service. Measuring service performance using a histogram to capture response times as a distribution may be appropriate. Extracting the 95th percentile of response time may be useful as well.
+
+There does not be a use case requiring consistency between values of different metrics
+
+
+Dali Use Cases
+==============
+
+From information gathered, Dali plans to keep counts and rates for many of the items it manages.
+
+**********************
+Framework Requirements
+**********************
+Not sure if we need this section or not. Might be good to collect high level instead of per area requirements.
+
+Metric Values
+===============
+Metric value requirements.  Capture the following:
+
+* intrinsic value ownership
+* types
+
+Reporting
+=========
+Cover how reports are generated, including how they are triggered.
+
+
+Sinks
+=====
+Document requirements of sinks to collect and report values to a backend collection systems
+
+Supported Collection Systems
+----------------------------
+Document the initial set of sinks. Probably the following:
+
+* Prometheus
+* Data dog
+* Elasticsearch
+* Null
```

Comment:
Do you mean file?  I'm not sure what a null sink means.

Reply:
I should add File to the list as well. I always saw the File sink as more of a development tool and not something that would be used in production. However, it appears that there may be a use for the file sink beyond development. A Null sink would simply allow for enabling the collection and reporting of metrics, but without them actually going anywhere. I see it being useful when duplicating environments and settings but you don't want to fully setup collection. Simply change the sink in the global part of the metrics config to the null sink. Everything runs normally, but the metrics are thrown away. I see this being useful for testing.